### ISSUE TITLE
Allow to use the dispatcher as a middleware

### ISSUE THREAD

**Issue by oscarotero** [relayphp/Relay.Relay#30]
(This comes from https://github.com/oscarotero/psr7-middlewares/issues/34)

It should be great to have the posibility of using the dispatcher as a middleware of other dispatcher. Example:

``` php
use Relay\Runner;

$group = new Runner([
    new Middleware1(),
    new Middleware2(),
]);

$dispatcher = new Runner([
    $group,
    new Middleware3()
]);
```

**Comment by Isinlor** [comment]
The question is do we overload Runner `__invoke` method or do we provide a wrapper? It may be a good idea to create a separate package that could take object with `function($request, $response) : Response` signature and change it to a middleware.

@edit I was to quick to answer - it is possible already with Relay 2.0. I just tested it. We can add unit test to make sure that this case will works in future. @pmjones What do you think?

**Comment by oscarotero** [comment]
I didn't test it but after a quick look in the code seems like this is not implemented (at least not directly in Runner class): https://github.com/relayphp/Relay.Relay/blob/2.x/src/Runner.php#L72

Providing an optional third argument with the $next callable would be enough.

**Comment by Isinlor** [comment]
Indeed, Runner `__invoke()` doesn't have `$next` argument in signature, but PHP allows to call a function with arguments that are not in the signature of a function. [Arguments are silently ignored in that case.](http://php.net/manual/en/functions.arguments.php#functions.variable-arg-list.old)

So it will work, but `Middleware3` from your example won't be called. This also lead to a question. Should `$next` in Runner be called before or after Runner queue or not called at all, as it is the case now.

This makes me think that Runner should not be responsible for handling `$next` and it should be delegated to a specialized middelware.

**Comment by oscarotero** [comment]
IMO, the workflow of the "sub-middleware" should respect the workflow of the main middleware. This means:
- Middleware1 is executed until the $next call
- Middleware2 is executed until the $next call
- Middleware3 is executed until the $next call
- Middleware3 is finished
- Middleware2 is finished
- Middleware1 is finished

**Comment by pmjones** [comment]
@oscarotero @Isinlor :

@mindplay-dk points out elsewhere that a middleware _dispatcher_ (which is what Runner is) need not have the same signature as the middleware it dispatches.

Perhaps the answer here is a DispatcherHandler middleware piece that follows the `fn (req, res, next)` signature? It could then be queued up like any other middleware.

Finally, if you want to send a PR against the 2.x branch to implement the dispatcher-as-middleware idea, I'll be happy to review it.

Thanks for your interest here!

**Comment by oscarotero** [comment]
Here's the PR #31 🎉

**Comment by pmjones** [comment]
Fixed by #31
